File: examples/stream.ml
let stream_of_string s = Stdlib.Stream.of_string s in
Callback.register "stream_of_string" stream_of_string;;

let stream_next s = Stdlib.Stream.next s |> Char.code |> string_of_int in
Callback.register "stream_next" stream_next;;

let stream_empty s = try Stdlib.Stream.empty s; "true" with Stdlib.Stream.Failure -> "false" in
Callback.register "stream_empty" stream_empty;;

File: examples/stream.rs
use caml_dyn_call::*;
use eyre::Result;

fn main() -> Result<()> {
    init(std::path::Path::new("examples/stream.ml"))?;

    let stream = dyn_call!("stream_of_string", "Hello, World!")?;
    loop {
        let c = dyn_call!("stream_empty", &stream)?;
        match c.get_str()?.as_str() {
            "true" => break,
            "false" => {
                let c = dyn_call!("stream_next", &stream)?;
                let c = c.get_str()?.parse()?;
                let c = char::from_u32(c).unwrap();
                print!("{}", c);
            }
            _ => unreachable!(),
        }
    }

    Ok(())
}
